### ISSUE TITLE
Returns "detail": "Authentication credentials were not provided."

### ISSUE THREAD

**Issue by binyammesfin** [marcelo225/django-keycloak-auth#19]
I have all the setup done. I can introspect using Postman. But when I try to access an API endpoint from my DRF, I get "detail": "Authentication credentials were not provided."

Does Django need to authenticate with Keycloak to introspect tokens? I am not sure what I am missing and any help is greatly appreciated. 

I am running both the API endpoint and keycloak via docker-compose. 

Thanks.

**Comment by binyammesfin** [comment]
I changed the config in settings.py to 'KEYCLOAK_SERVER_URL': 'http://keycloak:8080/auth' (the container name), and I can get a response, but it states "detail": "Invalid or expired token. Verify your Keycloak configuration." for a login that I can introspect. 

I am trying from Postman, not from the client yet. 

Thanks.

**Comment by marcelo225** [comment]
Hi, @binyammesfin !

Sorry about that, but have you already configured the Realm and Client on Keycloak yet? 
This is a configuration Keycloak that you have to do before you use the django-keycloak-auth.

You can read Keycloak documentation about how you can configure the Realm and Client before you use this lib.

I dropped the following URL to replicate in your Keycloak container:
- https://medium.com/keycloak/keycloak-realm-client-configuration-dfd7c8583489

When you've done this config, you can configure **django-keycloak-auth** like this:

```python
KEYCLOAK_EXEMPT_URIS = [] # Realtive URLs in your application, separeted by comma
KEYCLOAK_CONFIG = {
    'KEYCLOAK_SERVER_URL': 'http://localhost:8080/auth',
    'KEYCLOAK_REALM': 'Realm name',
    'KEYCLOAK_CLIENT_ID': 'Client ID',
    'KEYCLOAK_CLIENT_SECRET_KEY': 'Client Secret Key'
}
```
I hope it helps you!

Best regards!

**Comment by binyammesfin** [comment]
Hi @marcelo225 

Yes, I have done all and I can log in and log out from the frontend fine. The client receives a token, which I am passing via RTK Query's fetchBaseQuery to Django. I can also see the Token is included in the request header. 

I have followed your configuration to a T and I am getting a 401 response. 

In XHR, I also see the this:

Request URL: http://localhost:8080/auth/realms/bhh/protocol/openid-connect/token
Request Method: POST
Status Code: 200 OK
Remote Address: [::1]:8080
Referrer Policy: strict-origin-when-cross-origin

It only fails when it actually hits the view. If I remove the

    keycloak_roles = {
        'GET': ['bhh-admin',],
        'POST': ['admin', ],
        'UPDATE': ['admin', ],
        'DELETE': ['admin', ],
        'PATCH': ['admin',],
    }
from the view, the resource loads fine. 

One question, do I need to disable "AUTH_USER_MODEL" from Django now I am relying on Keycloak?

Here what the 401 Error details are:

**General**
Request URL: http://localhost:8000/api/region/
Request Method: GET
Status Code: 401 Unauthorized
Remote Address: [::1]:8000
Referrer Policy: strict-origin-when-cross-origin

**Response Headers**
Access-Control-Allow-Origin: http://localhost:3000
Content-Length: 75
Content-Type: application/json
Cross-Origin-Opener-Policy: same-origin
Date: Mon, 14 Feb 2022 15:23:17 GMT
Referrer-Policy: same-origin
Server: WSGIServer/0.2 CPython/3.10.2
Vary: Origin
X-Content-Type-Options: nosniff
X-Frame-Options: DENY

**Request Headers**
Accept: */*
Accept-Encoding: gzip, deflate, br
Accept-Language: en-US,en;q=0.9,la;q=0.8
authorization: Bearer eyJhbGciOiJSUzI1NiIsInR5cCIgOiAiSldUIiwia2lkIiA6ICJrSVNPR2hRRGZLWjZNZXg2SFRRcVluRkMzeEZTSlV6T3UyY3p3MVN0SDdBIn0.eyJleHAiOjE2NDQ4NTI0OTcsImlhdCI6MTY0NDg1MjE5NywiYXV0aF90aW1lIjoxNjQ0ODUxOTQ4LCJqdGkiOiJhY2RiZTI4YS01NjFjLTRjNjMtYWE3Yy1lMmU3MTYzZmY3NDAiLCJpc3MiOiJodHRwOi8vbG9jYWxob3N0OjgwODAvYXV0aC9yZWFsbXMvYmhoIiwiYXVkIjpbImJoaC1hcGkiLCJhY2NvdW50Il0sInN1YiI6ImU2ZTdjZDA2LWU5YWYtNDA1Ny1hZTRlLTZhZWY1YzE1ZjYyNCIsInR5cCI6IkJlYXJlciIsImF6cCI6ImJoaC1kYXNoYm9hcmQiLCJub25jZSI6ImViMDQyMzEyLTI2YzAtNGQyNy04MmE4LWU5NDA0ZDgxYmE0YyIsInNlc3Npb25fc3RhdGUiOiIyZGJkMDFkNy00MmJiLTQ5M2UtYWI3Zi03MGI1OWQ5MTVmNGUiLCJhY3IiOiIwIiwiYWxsb3dlZC1vcmlnaW5zIjpbImh0dHA6Ly9sb2NhbGhvc3Q6MzAwMCJdLCJyZWFsbV9hY2Nlc3MiOnsicm9sZXMiOlsiYmhoLWFkbWluIiwiYmhoLWFkbWluLWNsaWVudCIsImJoaC11c2VyIiwib2ZmbGluZV9hY2Nlc3MiLCJhcHAtYWRtaW4iLCJ1bWFfYXV0aG9yaXphdGlvbiIsImFwcC11c2VyIiwiZGVmYXVsdC1yb2xlcy1iaGgiXX0sInJlc291cmNlX2FjY2VzcyI6eyJiaGgtYXBpIjp7InJvbGVzIjpbImFkbWluIiwidXNlciJdfSwiYmhoLWRhc2hib2FyZCI6eyJyb2xlcyI6WyJhZG1pbiIsInVzZXIiXX0sImFjY291bnQiOnsicm9sZXMiOlsibWFuYWdlLWFjY291bnQiLCJtYW5hZ2UtYWNjb3VudC1saW5rcyIsInZpZXctcHJvZmlsZSJdfX0sInNjb3BlIjoib3BlbmlkIHByb2ZpbGUgY2xpZW50X3JvbGVzX2JoaC1kYXNoYm9hcmQgZW1haWwiLCJzaWQiOiIyZGJkMDFkNy00MmJiLTQ5M2UtYWI3Zi03MGI1OWQ5MTVmNGUiLCJlbWFpbF92ZXJpZmllZCI6dHJ1ZSwicm9sZXMiOlsiYWRtaW4iLCJ1c2VyIl0sIm5hbWUiOiJCaW55YW0gTWVzZmluIiwicHJlZmVycmVkX3VzZXJuYW1lIjoiYmlueWFtbWVzZmluIiwiZ2l2ZW5fbmFtZSI6IkJpbnlhbSIsImZhbWlseV9uYW1lIjoiTWVzZmluIiwiZW1haWwiOiJiaW55YW1tZXNmaW5AeWFob28uY29tIn0.T8FVb1MZeZyxxVX6TO2SlPoQhAox_P0yvC22mIQi_94GpDG9KZx2oxJYvJORGHHeeSF_mAGU4tX-FqxMH4XxN_oaM308j_iiXitCBPc5Qi6GMVP2Dt8gKRX6psXNIPXEKvs-BVuCRhpkVvheOKCm2ORKriYB1Sw6j4RWe_-ZGIRgltNUWKuW-eVJsafQfqLnF6MOVGtWjWwxrgl76nv0R5Swr_pWAGw7XmH38nbfsVIYJ29EprStEFkKuwAJQBwzxnQLmXW745yTwMZzS8S655jy_9YBTyS27w4B5ic5pNeli8bY3Kbi6uLPQACaU9SNJHlaJuvL_vkXKADr_59Q-w
Connection: keep-alive
Host: localhost:8000
Origin: http://localhost:3000
Referer: http://localhost:3000/
sec-ch-ua: " Not A;Brand";v="99", "Chromium";v="98", "Google Chrome";v="98"
sec-ch-ua-mobile: ?0
sec-ch-ua-platform: "Windows"
Sec-Fetch-Dest: empty
Sec-Fetch-Mode: cors
Sec-Fetch-Site: same-site
User-Agent: Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/98.0.4758.82 Safari/537.36

Thanks so much and appolgies if I missed something obvious!

Best regards!

**Comment by binyammesfin** [comment]
Sorry for the multiple posts, but the issue is that the middleware thinks the token is not active and throws "{detail: "Invalid or expired token. Verify your Keycloak configuration."}" I check the configuration where

SSO Session Idle is 30  min 
SSO Session Max is 10 hours
Access Token Lifespan is 5 minutes

Thanks

**Comment by binyammesfin** [comment]
Hi @marcelo225, 

Thank you for the help and I finally managed to solve it:

There were two issues:

1. The localhost in the KEYCLOAK_SERVER_URL': 'http://localhost:8080/auth' was not properly reached and I had to implement host.docker.internal in the docker-compose file to provide an alternative ('KEYCLOAK_SERVER_URL': 'http://host.docker.internal:8080/auth',).
2. Once that was changed, I had to also update the frontend configuration to the same address (which was also localhost:8080) to match the host.docker.internal address. I now know if the token was generated from a different address, keycloak does not authenticate the user properly. 

I hope this will help someone. 

Thank you for the amazing package. 

Best regards,
